File: backend/app/api/v1/nodes.py
"""节点管理 API(admin 端,走 session auth)— `/api/v1/nodes/*`。

MVP-1 远程 agent,设计稿:`进度/设计/远程VPS脚本执行调研.md` § 5 / § 7。

端点清单(6 个)
---------------
- GET    /nodes                         🔒 列表(含 online 派生)
- POST   /nodes                         🔒 创建,返回**一次性** 明文 token
- GET    /nodes/{id}                    🔒 详情
- PATCH  /nodes/{id}                    🔒 改 name/description/enabled
- POST   /nodes/{id}/regenerate-token   🔒 重新生成 token
- DELETE /nodes/{id}                    🔒 删除(local 不可删;有 instance 关联时拒)

业务异常统一抛 ``app.core.exceptions.*``,由 ``error_handler`` 转标准响应。
"""
from __future__ import annotations

from fastapi import APIRouter, Response, status

from app.db.models.node import Node
from app.deps import CurrentUser, DBSession
from app.schemas.node import (
    NodeCreate,
    NodeCreateResponse,
    NodeDetail,
    NodeListItem,
    NodeListResponse,
    NodeTokenResponse,
    NodeUpdate,
)
from app.services import node_service

router = APIRouter(prefix="/nodes", tags=["nodes"])


# ============================================================
# 序列化辅助
# ============================================================
def _node_to_item(node: Node) -> NodeListItem:
    """Node ORM → NodeListItem(含 online 派生 + metadata 反序列化)。"""
    return NodeListItem(
        id=node.id,
        slug=node.slug,
        name=node.name,
        description=node.description,
        is_local=bool(node.is_local),
        last_seen_at=node.last_seen_at,
        version=node.version,
        metadata=node_service.node_metadata_dict(node),
        enabled=bool(node.enabled),
        online=node_service.is_node_online(node),
        created_at=node.created_at,
        updated_at=node.updated_at,
    )


def _node_to_detail(node: Node) -> NodeDetail:
    item = _node_to_item(node)
    return NodeDetail(
        **item.model_dump(),
        deployed_scripts=node_service.node_deployed_scripts(node),
        pending_actions=node_service.node_pending_actions(node),
    )


# ============================================================
# 列表
# ============================================================
@router.get(
    "",
    response_model=NodeListResponse,
    summary="节点列表(含 online 派生状态)",
)
def list_nodes(
    db: DBSession,
    _user: CurrentUser,
) -> NodeListResponse:
    """🔒 列表(按 id 升序;local 永远在最前)。"""
    nodes = node_service.list_nodes(db)
    items = [_node_to_item(n) for n in nodes]
    return NodeListResponse(items=items, total=len(items))


# ============================================================
# 创建
# ============================================================
@router.post(
    "",
    response_model=NodeCreateResponse,
    status_code=status.HTTP_201_CREATED,
    summary="创建节点(返回一次性明文 token)",
)
def create_node(
    payload: NodeCreate,
    db: DBSession,
    _user: CurrentUser,
) -> NodeCreateResponse:
    """🔒 创建一个新远程 agent 节点。

    **重要**:响应里的 ``token`` 字段是明文,**仅此一次返回**。
    前端必须显示给用户保存,关闭弹窗后无法再取(DB 只存 bcrypt hash)。
    丢失只能 ``POST /nodes/{id}/regenerate-token``。
    """
    node, token = node_service.create_node(
        db,
        slug=payload.slug,
        name=payload.name,
        description=payload.description,
    )
    db.commit()
    return NodeCreateResponse(
        node=_node_to_detail(node),
        token=token,
    )


# ============================================================
# 详情
# ============================================================
@router.get(
    "/{node_id}",
    response_model=NodeDetail,
    summary="节点详情",
)
def get_node(
    node_id: int,
    db: DBSession,
    _user: CurrentUser,
) -> NodeDetail:
    """🔒 详情。"""
    node = node_service.get_node_by_id(db, node_id)
    return _node_to_detail(node)


# ============================================================
# 更新
# ============================================================
@router.patch(
    "/{node_id}",
    response_model=NodeDetail,
    summary="更新节点(name/description/enabled)",
)
def update_node(
    node_id: int,
    payload: NodeUpdate,
    db: DBSession,
    _user: CurrentUser,
) -> NodeDetail:
    """🔒 部分更新。

    限制:本地节点 enabled 不可改为 False。
    """
    node = node_service.update_node(
        db,
        node_id,
        name=payload.name,
        description=payload.description,
        enabled=payload.enabled,
    )
    db.commit()
    return _node_to_detail(node)


# ============================================================
# 重新生成 token
# ============================================================
@router.post(
    "/{node_id}/regenerate-token",
    response_model=NodeTokenResponse,
    summary="重新生成节点 token(旧 token 立即失效)",
)
def regenerate_token(
    node_id: int,
    db: DBSession,
    _user: CurrentUser,
) -> NodeTokenResponse:
    """🔒 旧 token 立即失效;新明文 token 仅此一次返回。"""
    token = node_service.regenerate_token(db, node_id)
    db.commit()
    return NodeTokenResponse(token=token)


# ============================================================
# 删除
# ============================================================
@router.delete(
    "/{node_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="删除节点(local 不可删;有 instance 时拒)",
)
def delete_node(
    node_id: int,
    response: Response,
    db: DBSession,
    _user: CurrentUser,
) -> Response:
    """🔒 删除节点。

    限制:
    - 本地节点(is_local=1)永远不可删
    - 有 instance 关联时拒绝(用户必须先迁移 instance.node_id 到其它节点)
    """
    node_service.delete_node(db, node_id)
    db.commit()
    response.status_code = status.HTTP_204_NO_CONTENT
    return response


# ============================================================
# 卸载节点上已部署的脚本(下发 pending_actions.delete)
# ============================================================
@router.post(
    "/{node_id}/deployed-scripts/{slug}/uninstall",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="卸载节点上已部署的脚本(下发删除指令,agent 拉取后 rm)",
)
def uninstall_node_script(
    node_id: int,
    slug: str,
    response: Response,
    db: DBSession,
    _user: CurrentUser,
) -> Response:
    """🔒 把 slug 加入节点 ``pending_actions.delete``。

    agent 下次 poll 拉到后删本地 ``scripts/<slug>/``,再 inventory-report ack →
    主面板从 ``deployed_scripts`` 移除。本地节点不适用(脚本在主面板,走 ``/scripts`` 删)。
    """
    node_service.request_uninstall_script(db, node_id, slug)
    db.commit()
    response.status_code = status.HTTP_204_NO_CONTENT
    return response

File: frontend/src/api/hooks/nodes.ts
/**
 * Nodes 数据 hooks(MVP-1 远程 agent)
 *
 * 后端 API:`backend/app/api/v1/nodes.py`
 * 设计稿:`进度/设计/远程VPS脚本执行调研.md` § 7
 *
 * Query keys:
 *   ['nodes']                    — root,创建/删除/regenerate 后 invalidate
 *   ['nodes', 'list']            — 列表
 *   ['nodes', 'detail', id]      — 详情
 *
 * 注意:
 *   - schema.d.ts 还没有 nodes 类型,本文件用 apiClient.GET<any> + 手写 TS 接口
 *   - create / regenerate-token 返回明文 token,**只展示一次**,前端组件要 modal 显示给用户复制
 */
import {
  useMutation,
  useQuery,
  useQueryClient,
  type UseMutationResult,
  type UseQueryResult,
} from '@tanstack/react-query';
import { toast } from 'sonner';

import { apiClient } from '@/api/client';
import { formatError } from '@/lib/error';

// ====================== 类型 ======================

export interface NodeListItem {
  id: number;
  slug: string;
  name: string | null;
  description: string | null;
  is_local: boolean;
  last_seen_at: string | null;
  version: string | null;
  metadata: Record<string, unknown>;
  enabled: boolean;
  online: boolean;
  created_at: string;
  updated_at: string;
}

export interface NodeDetail extends NodeListItem {
  /** agent 上报的已部署脚本 {slug: {sha256, deployed_at}} */
  deployed_scripts?: Record<string, { sha256?: string; deployed_at?: string }>;
  /** 待 agent 执行的指令(sync 拉取 / delete 删除) */
  pending_actions?: { sync?: string[]; delete?: string[] };
}

export interface NodeListResponse {
  items: NodeListItem[];
  total: number;
}

export interface NodeCreate {
  slug: string;
  name?: string | null;
  description?: string | null;
}

export interface NodeUpdate {
  name?: string | null;
  description?: string | null;
  enabled?: boolean;
}

/** 创建响应 — 含一次性明文 token */
export interface NodeCreateResponse {
  node: NodeDetail;
  token: string;
}

export interface NodeTokenResponse {
  token: string;
}

// ====================== Fetch helpers ======================

async function fetchNodes(signal?: AbortSignal): Promise<NodeListResponse> {
  const { data, error, response } = await apiClient.GET(
    '/api/v1/nodes' as never,
    { signal } as never,
  );
  if (error) {
    throw Object.assign(new Error(formatError(error)), {
      status: response?.status,
      detail: error,
    });
  }
  const raw = (data ?? { items: [], total: 0 }) as Partial<NodeListResponse>;
  return {
    items: Array.isArray(raw.items) ? raw.items : [],
    total: raw.total ?? 0,
  };
}

async function fetchNodeDetail(id: number, signal?: AbortSignal): Promise<NodeDetail> {
  const { data, error, response } = await apiClient.GET(
    `/api/v1/nodes/${id}` as never,
    { signal } as never,
  );
  if (error) {
    throw Object.assign(new Error(formatError(error)), {
      status: response?.status,
      detail: error,
    });
  }
  return data as unknown as NodeDetail;
}

async function createNode(payload: NodeCreate): Promise<NodeCreateResponse> {
  const { data, error, response } = await apiClient.POST(
    '/api/v1/nodes' as never,
    { body: payload } as never,
  );
  if (error) {
    throw Object.assign(new Error(formatError(error)), {
      status: response?.status,
      detail: error,
    });
  }
  return data as unknown as NodeCreateResponse;
}

async function updateNode(id: number, payload: NodeUpdate): Promise<NodeDetail> {
  const { data, error, response } = await apiClient.PATCH(
    `/api/v1/nodes/${id}` as never,
    { body: payload } as never,
  );
  if (error) {
    throw Object.assign(new Error(formatError(error)), {
      status: response?.status,
      detail: error,
    });
  }
  return data as unknown as NodeDetail;
}

async function deleteNode(id: number): Promise<void> {
  const { error, response } = await apiClient.DELETE(
    `/api/v1/nodes/${id}` as never,
    {} as never,
  );
  if (error) {
    throw Object.assign(new Error(formatError(error)), {
      status: response?.status,
      detail: error,
    });
  }
}

async function regenerateToken(id: number): Promise<NodeTokenResponse> {
  const { data, error, response } = await apiClient.POST(
    `/api/v1/nodes/${id}/regenerate-token` as never,
    {} as never,
  );
  if (error) {
    throw Object.assign(new Error(formatError(error)), {
      status: response?.status,
      detail: error,
    });
  }
  return data as unknown as NodeTokenResponse;
}

async function uninstallNodeScript(nodeId: number, slug: string): Promise<void> {
  const { error, response } = await apiClient.POST(
    `/api/v1/nodes/${nodeId}/deployed-scripts/${encodeURIComponent(slug)}/uninstall` as never,
    {} as never,
  );
  if (error) {
    throw Object.assign(new Error(formatError(error)), {
      status: response?.status,
      detail: error,
    });
  }
}

// ====================== Query keys ======================

export const nodesQueryKeys = {
  all: ['nodes'] as const,
  list: () => ['nodes', 'list'] as const,
  detail: (id: number) => ['nodes', 'detail', id] as const,
};

// ====================== Hooks ======================

/** GET /api/v1/nodes — 节点列表(含 online 派生字段) */
export function useNodes(): UseQueryResult<NodeListResponse, Error> {
  return useQuery({
    queryKey: nodesQueryKeys.list(),
    queryFn: ({ signal }) => fetchNodes(signal),
    staleTime: 15_000,
    refetchInterval: 30_000, // 每 30s 刷新 online 状态
  });
}

/** GET /api/v1/nodes/{id} — 节点详情 */
export function useNode(id: number | undefined): UseQueryResult<NodeDetail, Error> {
  return useQuery({
    queryKey: nodesQueryKeys.detail(id ?? 0),
    queryFn: ({ signal }) => fetchNodeDetail(id as number, signal),
    enabled: !!id,
    staleTime: 15_000,
  });
}

/** POST /api/v1/nodes — 创建节点(返回一次性明文 token) */
export function useCreateNode(): UseMutationResult<NodeCreateResponse, Error, NodeCreate> {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (payload) => createNode(payload),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: nodesQueryKeys.all });
      // 不直接 toast(由 NodeList Dialog 显示 token);避免 toast 闪一下盖住关键信息
    },
  });
}

/** PATCH /api/v1/nodes/{id} — 改 name/description/enabled */
export function useUpdateNode(): UseMutationResult<NodeDetail, Error, { id: number; payload: NodeUpdate }> {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ id, payload }) => updateNode(id, payload),
    onSuccess: (_, { id }) => {
      void qc.invalidateQueries({ queryKey: nodesQueryKeys.all });
      void qc.invalidateQueries({ queryKey: nodesQueryKeys.detail(id) });
      toast.success('节点已更新');
    },
  });
}

/** DELETE /api/v1/nodes/{id} */
export function useDeleteNode(): UseMutationResult<void, Error, number> {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (id) => deleteNode(id),
    onSuccess: (_, id) => {
      void qc.invalidateQueries({ queryKey: nodesQueryKeys.all });
      qc.removeQueries({ queryKey: nodesQueryKeys.detail(id) });
      toast.success('节点已删除');
    },
  });
}

/** POST /api/v1/nodes/{id}/regenerate-token */
export function useRegenerateNodeToken(): UseMutationResult<NodeTokenResponse, Error, number> {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (id) => regenerateToken(id),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: nodesQueryKeys.all });
      // token 由 NodeList Dialog 显示
    },
  });
}

/** POST /api/v1/nodes/{id}/deployed-scripts/{slug}/uninstall — 下发删除指令给 agent */
export function useUninstallNodeScript(): UseMutationResult<
  void,
  Error,
  { nodeId: number; slug: string }
> {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ nodeId, slug }) => uninstallNodeScript(nodeId, slug),
    onSuccess: (_, { nodeId }) => {
      void qc.invalidateQueries({ queryKey: nodesQueryKeys.detail(nodeId) });
      void qc.invalidateQueries({ queryKey: nodesQueryKeys.all });
      toast.success('已下发删除指令,agent 下次拉取(最长 30s)时删除并回报');
    },
  });
}
